src/Nethermind/Nethermind.Blockchain/Processing/ProducingBlockTransactionProcessingStrategy.cs

```diff
@@ -0,0 +1,90 @@
+// 
+//  You should have received a copy of the GNU Lesser General Public License
+//  along with the Nethermind. If not, see <http://www.gnu.org/licenses/>.
+// 
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Nethermind.Core;
+using Nethermind.Core.Collections;
+using Nethermind.Core.Specs;
+using Nethermind.Evm;
+using Nethermind.Evm.Tracing;
+using Nethermind.State;
+using Nethermind.State.Proofs;
+using Nethermind.TxPool;
+
+namespace Nethermind.Blockchain.Processing
+{
+    public class ProducingBlockTransactionProcessingStrategy : ITransactionProcessingStrategy
+    {
+        private readonly ITransactionProcessor _transactionProcessor;
+        private readonly IStateProvider _stateProvider;
+        private readonly IStorageProvider _storageProvider;
+        private readonly ProcessingOptions _options;
+        
+        public ProducingBlockTransactionProcessingStrategy(
+            ITransactionProcessor transactionProcessor, 
+            IStateProvider stateProvider,
+            IStorageProvider storageProvider, 
+            ProcessingOptions options)
+        {
+            _transactionProcessor = transactionProcessor;
+            _stateProvider = stateProvider;
+            _storageProvider = storageProvider;
+            _options = options;
+        }
+        
+        public TxReceipt[] ProcessTransactions(Block block, ProcessingOptions processingOptions, IBlockTracer blockTracer, BlockReceiptsTracer receiptsTracer, IReleaseSpec spec, EventHandler<TxProcessedEventArgs> TransactionProcessed)
+        {
+            IEnumerable<Transaction> transactions = block.GetTransactions(out _);
+
+            int i = 0;
+            LinkedHashSet<Transaction> transactionsForBlock = new(DistinctCompareTx.Instance);
+            foreach (Transaction currentTx in transactions)
+            {
+                if (!transactionsForBlock.Contains(currentTx))
+                {
+                    // No more gas available in block
+                    if (currentTx.GasLimit > block.Header.GasLimit - block.GasUsed)
+                    {
+                        break;
+                    }
+
+                    ProcessTransaction(block, currentTx, i++, receiptsTracer, TransactionProcessed);
+                    transactionsForBlock.Add(currentTx);
+                }
+            }
+            block.TrySetTransactions(transactionsForBlock.ToArray());
+            block.Header.TxRoot = new TxTrie(block.Transactions).RootHash;
+            return receiptsTracer.TxReceipts!;
+        }
+        
+        private void ProcessTransaction(Block block, Transaction currentTx, int index, BlockReceiptsTracer receiptsTracer, EventHandler<TxProcessedEventArgs> TransactionProcessed)
+        {
+            if ((_options & ProcessingOptions.DoNotVerifyNonce) != 0)
+            {
+                currentTx.Nonce = _stateProvider.GetNonce(currentTx.SenderAddress);
+            }
```

Comment:
Can we remove this from ProcessingOptions and not pass processing options here at all, but when using tracing use different strategy?

Reply:
This is in scope of next PR